### DISCUSSION TITLE
How to handle HTTP_PUT?

### DISCUSSION THREAD

**Discussion by Levak** [mathieucarbou/ESPAsyncWebServer#158] [Q&A] [answered]
Hi,
I am porting to `ESPAsyncWebServer` a codebase that worked with the `Webserver` lib.
One of the requests had method type `HTTP_PUT`. I tried to find in documentation and in the source code any mention of `PUT` as a method type, but I cannot wrap my head around it.

Here is a link to the existing code:
https://github.com/tiredboffin/sdwifi/blob/2f045716dc03dca25dccdf4d0004d5a2209a4665/sdwifi.ino#L751-L771

From what I understand, `ArUploadHandlerFunction` is only for multipart `POST` requests, while `PUT` requests send a continuous stream for one file. `ArBodyHandlerFunction` is also tagged to be `POST`-only.

Which leaves us to only use `ArRequestHandlerFunction`.
Any hints where to start ?

```c
void init()
{
  server.on("^\\/upload\\/([^/]+)$", HTTP_PUT, handleUploadPUT);
}
void handleUploadPUT(AsyncWebServerRequest *request)
{
  String path = request->pathArg(0);
  [...]
}
```

**Comment by mathieucarbou** [comment]
> From what I understand, `ArUploadHandlerFunction` is only for multipart `POST` requests, while `PUT` requests send a continuous stream for one file. `ArBodyHandlerFunction` is also tagged to be `POST`-only.

Where do you see that ?

I saw no such thing in the code, plus, the HTTP spec supports a body, even for GET requests strangely.

**Comment by Levak** [comment]
[[EDIT: I was writting this answer 10min after the OP. But I had troubles validating that it works. ]]

I guess I misread the source code of `ESPAsyncWebServer`. `ArBodyHandlerFunction` is used when NOT a POST Plain text :D 

Here is a simple snippet for future references:
⚠️ Important: define `ASYNCWEBSERVER_REGEX` in build properties (in Arduino IDE, my trick is to add it to `pins_arduino.h`), otherwise, the regex won't be effective.

```c
#include <ESPAsyncWebServer.h>

void init() {
  server.on("^/upload/(.+)$", HTTP_PUT, handleUpload, nullptr, handleUploadPUT);
}

struct HandlePUTStates {
  File dataFile;
};

void handleUploadPUT(AsyncWebServerRequest *request, uint8_t* data, size_t len, size_t index, size_t total)
{
  String path = request->pathArg(0);
  struct HandlePUTStates * states;

  if (index == 0) {
    states = new HandlePUTStates();
    states->dataFile = File();
    request->_tempObject = states;

    File dataFile = fileSystem.open(path, FILE_WRITE);

    states->dataFile = dataFile;
  } else {
    states = (struct HandlePUTStates *) request->_tempObject;
  }

  states->dataFile.write(data, len);

  if(index + len == total) {
    states->dataFile.close();
    states->dataFile = File();
  }
}
```

I am just missing this piece:
```c
  case RAW_ABORTED:
    dataFile.close();
    fileSystem.remove(path);
    log_v("Upload PUT: file aborted");
    break;
```
I did not find yet a way to detect a request abort.

**Reply by mathieucarbou** [reply, marked as answer]
> I did not find yet a way to detect a request abort.

aborting a request from client-side would close the connection thus trigger the disconnect clalback.

did you try `request->onDisconnect([]() { ... })` ?

Never tried, but it is implemented.

**Reply by Levak** [reply]
Great!
That does the trick. Inside the `if (index == 0)` I added:
```c
    request->onDisconnect([states, path](){
      states->dataFile.close();
      states->dataFile = File();
      fileSystem.remove(path);
    });
```

And after testing, it works fine. a `/list` returns 404 if upload was aborted, 200 if upload was successful. 

Have a good one 👍
